### PR TITLE
Refactor Client class

### PR DESCRIPTION
**Pull request by Seb-stian**
Notable changes:
- fixed some (not all) nullable warnings
- uses Socket directly, instead of TcpClient
- fixes a bug where client can't connect again after disconnecting

Since there are a lot of small changes, you might prefer to review the changed file directly [here](https://github.com/ObsidianMC/Obsidian/blob/client-refactor/Obsidian/Client.cs).

### CODE REVIEW

File: Obsidian/Entities/Player.cs [hunk 1]
```diff
@@ -880,6 +880,57 @@ private async Task PickupNearbyItemsAsync(float distance = 0.5f)
         }
     }
 
+    internal async Task UpdateChunksAsync(bool unloadAll = false)
```
Comment (Jonpro03) [inline Obsidian/Entities/Player.cs:883]: I would argue that updating chunks is Client and not Player, but I think @Tides agrees with you too so majority rules.

**Review by Jonpro03 — APPROVED** [review]
Yeah! This looks good

File: Obsidian/Entities/Player.cs (hunk 1 above)
Comment (Naamloos) [reply Obsidian/Entities/Player.cs:883]: fun fact: I personally agree on chunks being a client thing.

**Review by Naamloos — COMMENTED** [review]
Some discussion needed

**Review by Naamloos — APPROVED** [review]

**Pull request merged**